### DISCUSSION TITLE
Asynchronous model for AsyncImage

### DISCUSSION THREAD

**Discussion by kurthuwig** [coil-kt/coil#2643] [Q&A]
I have a URL that requires an OAuth access token that might need a refresh. So the generation of the request can only be done asynchronously. How can I do this, so that I do not have to use my own placeholder but instead the one that AsyncImage already provides? According to the docs, I have to provide the finished URL as a model.

I imagine something like this:

        AsyncImage(
            modifier = modifier,
            model = { ... code that provides a URL ... },
            contentDescription = contentDescription,
            contentScale = ContentScale.Crop,
        )

I've tried it, but the code never gets executed - which means it is possible to add this feature :-)

**Comment by colinrtwhite** [comment]
Your best bet is to create a custom fetcher that performs the network request then delegates to the network fetcher. I added a section to the docs about this [here](https://coil-kt.github.io/coil/image_pipeline/).

Using a function as a `model` param would not work as we can't compute a valid memory cache key for it.
